src/evaluation/evaluator.h

```diff
@@ -614,43 +572,428 @@ class Evaluator {
         m_ply--;
     }
 
-    void startTimeHandler()
-    {
-        bool wasStopped = m_isStopped.exchange(false);
-        if (!wasStopped) {
-            return; /* nothing to do here */
-        }
+    static uint8_t s_numSearchers;
+    static uint64_t s_nodes;
 
-        std::thread timeHandler([this] {
-            while (!m_isStopped) {
-                {
-                    std::scoped_lock lock(m_timeHandleMutex);
-                    if (m_isStopped)
-                        return; /* stopped elsewhere */
+    uint8_t m_searcherId {};
+    uint8_t m_ply {};
+    Repetition m_repetition;
+    MoveOrdering m_moveOrdering {};
+    uint8_t m_selDepth {};
+    std::optional<movegen::Move> m_ttMove {};
 
-                    if (std::chrono::system_clock::now() > m_endTime) {
-                        m_isStopped = true;
-                    }
-                }
+    syzygy::WdlResult m_wdl
+        = syzygy::WdlResultTableNotActive;
+    uint8_t m_dtz {};
 
-                std::this_thread::sleep_for(std::chrono::milliseconds(1));
+    BitBoard m_board;
+    uint64_t m_hash;
+
+    std::promise<SearcherResult> m_searchPromise;
+    std::future<SearcherResult> m_futureResult;
+
+    /* search configs */
+    constexpr static inline uint32_t s_fullDepthMove { 4 };
+    constexpr static inline uint32_t s_reductionLimit { 3 };
+    constexpr static inline int32_t s_futilityMargin { 100 };
+    constexpr static inline int32_t s_futilityEvaluationMargin { 120 };
+    constexpr static inline int32_t s_razorMarginShallow { 125 };
+    constexpr static inline int32_t s_razorMarginDeep { 175 };
+    constexpr static inline uint8_t s_razorDeepReductionLimit { 2 };
+
+    constexpr static inline uint8_t s_nullMoveReduction { 2 };
+};
+
+uint8_t Searcher::s_numSearchers { 0 };
+uint64_t Searcher::s_nodes { 0 };
+
+std::atomic_bool Searcher::s_searchStopped { true };
+std::atomic_bool Searcher::s_killed { false };
+
+class Evaluator {
+public:
+    Evaluator()
+    {
+        resizeSearchers(1);
+    };
+
+    void resizeSearchers(uint8_t size)
+    {
+        if (size == m_searchers.size()) {
+            return;
+        }
+
+        else {
+            for (uint8_t i = m_searchers.size(); i < size; i++) {
+                m_searchers.emplace_back(std::make_unique<Searcher>());
+            }
+        }
+
+        m_searchers.resize(size);
+        m_threadPool.resize(size + 2);
```

Comment:
I think we can simplify this a fair bit (it's non performance critical - will be done from uci options and not engine loop):

```suggestion
        if (size == m_searchers.size()) {
            return;
        }

	m_searchers.clear();
	m_searchers.reserve(size);
	
        for (uint8_t i = m_searchers.size(); i < size; i++) {
            m_searchers.emplace_back(std::make_unique<Searcher>());
        }

	/* add main thread + time handler as well */
        m_threadPool.resize(size + 2);
```

Reply:
How is this a simplification? Do you mean in terms of understand the execution?

Are there not any states we want to keep persistent when the number of threads is changed?

Reply:
The thread size is only set at start and should, in theory, never be changed "runtime".
In any case, we do not want to persist any state. If it's done mid game, it will be super fucked anyways ^^

Therefore I think it's important to just keep it as simple as possible and not try to "optimize for something that won't happen" :)